File: typescript/0143-reorder-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 Do not return anything, modify head in-place instead.
 */
function reorderList(head: ListNode | null): void {
    let slow: ListNode | null | undefined = head;
    let fast: ListNode | null | undefined = head?.next;

    while (fast && fast.next) {
        slow = slow?.next;
        fast = fast?.next?.next;
    }

    // reverse second half
    if (!(slow && slow.next)) return;
    let second: ListNode | null = slow.next;
    slow.next = null;
    let prev: ListNode | null = null;
    while (second) {
        let temp = second.next;
        second.next = prev;
        prev = second;
        second = temp;
    }

    // merge two halfs
    let first: any = head;
    second = prev;
    while (second) {
        let temp1 = first.next;
        let temp2 = second.next;
        first.next = second;
        second.next = temp1;
        first = temp1;
        second = temp2;
    }
}

File: go/0143-reorder-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reorderList(head *ListNode)  {
    slow := head
    fast := head.Next
    
    for fast != nil && fast.Next != nil {
        fast = fast.Next.Next
        slow = slow.Next
    }
    
    reversed := reverse(slow.Next)
    slow.Next = nil
    
    curr := head
    
    for curr != nil && reversed != nil {
        next := curr.Next
        revNext := reversed.Next
        curr.Next = reversed
        reversed.Next = next
        curr = next
        reversed = revNext
    }
}

func reverse(node *ListNode) *ListNode {
    var prev, curr *ListNode = nil, node
    
    for curr != nil {
        next := curr.Next
        curr.Next = prev
        prev = curr
        curr = next
    }
    
    return prev
}